File: js/commands/text.js
/* ROOT OF CRIME – text commands + pager */

Object.assign(Terminal, {
  _cmdCat(args, stdin) {
    if (!args.length) return this._ok(stdin || '');
    const chunks = [];
    for (const a of args) {
      const got = this._readFile(a);
      if (got.error) return this._err(got.error);
      chunks.push(got.text);
      if (this.ctx) {
        this.ctx.readFiles = this.ctx.readFiles || [];
        if (!this.ctx.readFiles.includes(got.path)) this.ctx.readFiles.push(got.path);
      }
    }
    return this._ok(chunks.join(''));
  },

  _cmdLess(args, stdin) {
    const paths = args.filter((a) => !a.startsWith('-'));
    let text = '';
    let title = '';
    if (paths.length) {
      const chunks = [];
      for (const a of paths) {
        const got = this._readFile(a);
        if (got.error) return this._err(got.error.replace(/^cat:/, 'less:'));
        chunks.push(got.text);
        if (this.ctx) {
          this.ctx.readFiles = this.ctx.readFiles || [];
          if (!this.ctx.readFiles.includes(got.path)) this.ctx.readFiles.push(got.path);
        }
        title = got.path;
      }
      text = chunks.join('\n');
      if (paths.length > 1) title = paths.length + ' files';
    } else {
      text = stdin || '';
      title = '(stdin)';
      if (!text) return this._err('less: missing filename (try: less FILE)');
    }
    const lines = String(text).replace(/\n$/, '').split('\n');
    this._pager = {
      lines,
      pos: 0,
      title,
      saved: this.outputEl ? this.outputEl.innerHTML : ''
    };
    this._renderPager();
    return { stdout: '', stderr: '', code: 0 };
  },

  _pagerPageSize() {
    const el = this.screenEl || this.outputEl;
    if (!el || !el.clientHeight) return 20;
    let lh = 18;
    try {
      const style = window.getComputedStyle(el);
      const parsed = parseFloat(style.lineHeight);
      if (parsed) lh = parsed;
    } catch {
      /* headless */
    }
    return Math.max(8, Math.floor(el.clientHeight / lh) - 1);
  },

  _renderPager() {
    if (!this._pager || !this.outputEl) return;
    const page = this._pagerPageSize();
    const { lines, pos, title } = this._pager;
    const slice = lines.slice(pos, pos + page);
    this.outputEl.innerHTML = this._escape(slice.join('\n')) + '\n';
    const atEnd = pos + page >= lines.length;
    const pct = lines.length ? Math.min(100, Math.round(((pos + slice.length) / lines.length) * 100)) : 100;
    const label = atEnd ? `--END-- ${title}` : `--More--(${pct}%) ${title}`;
    if (this.promptEl) this.promptEl.textContent = label;
    if (this.inputEl) {
      this.inputEl.value = '';
      this.inputEl.readOnly = true;
      this.inputEl.placeholder = '';
    }
    this._scroll();
  },

  _pagerPage(dir) {
    if (!this._pager) return;
    const page = this._pagerPageSize();
    const max = Math.max(0, this._pager.lines.length - page);
    this._pager.pos = Math.max(0, Math.min(max, this._pager.pos + dir * page));
    this._renderPager();
  },

  _pagerLine(dir) {
    if (!this._pager) return;
    const page = this._pagerPageSize();
    const max = Math.max(0, this._pager.lines.length - page);
    this._pager.pos = Math.max(0, Math.min(max, this._pager.pos + dir));
    this._renderPager();
  },

  _quitPager() {
    if (!this._pager) return;
    const { lines, pos, saved } = this._pager;
    const page = this._pagerPageSize();
    const shown = lines.slice(pos, pos + page).join('\n');
    this._pager = null;
    if (this.outputEl) {
      this.outputEl.innerHTML = (saved != null ? saved : '') + this._escape(shown) + (shown ? '\n' : '');
    }
    if (this.inputEl) {
      this.inputEl.readOnly = false;
      this.inputEl.placeholder = '';
      this.inputEl.value = '';
    }
    this._updatePrompt();
    this._scroll();
  },

  _onPagerKey(e) {
    e.preventDefault();
    e.stopPropagation();
    const k = e.key;
    if (k === 'q' || k === 'Q' || k === 'Escape') this._quitPager();
    else if (k === ' ' || k === 'f' || k === 'F' || k === 'PageDown') this._pagerPage(1);
    else if (k === 'b' || k === 'B' || k === 'PageUp') this._pagerPage(-1);
    else if (k === 'Enter' || k === 'j' || k === 'ArrowDown') this._pagerLine(1);
    else if (k === 'k' || k === 'ArrowUp') this._pagerLine(-1);
    else if (k === 'g') {
      this._pager.pos = 0;
      this._renderPager();
    } else if (k === 'G') {
      this._pager.pos = Math.max(0, this._pager.lines.length - this._pagerPageSize());
      this._renderPager();
    } else if (k === 'c' && e.ctrlKey) {
      this._quitPager();
    }
  },

  _cmdHeadTail(args, stdin, kind) {
    let n = 10;
    const paths = [];
    for (let i = 0; i < args.length; i++) {
      if (args[i] === '-n' && args[i + 1]) {
        n = parseInt(args[i + 1], 10) || 10;
        i++;
      } else if (/^-n?\d+$/.test(args[i])) {
        n = parseInt(args[i].replace('-n', '').replace('-', ''), 10) || 10;
      } else paths.push(args[i]);
    }
    let text = stdin;
    if (paths.length) {
      const got = this._readFile(paths[0]);
      if (got.error) return this._err(got.error.replace(/^cat/, kind));
      text = got.text;
      if (this.ctx) {
        this.ctx.readFiles = this.ctx.readFiles || [];
        if (!this.ctx.readFiles.includes(got.path)) this.ctx.readFiles.push(got.path);
      }
    }
    const lines = String(text || '').replace(/\n$/, '').split('\n');
    const slice = kind === 'head' ? lines.slice(0, n) : lines.slice(-n);
    return this._ok(slice.join('\n'));
  },

  _cmdWc(args, stdin) {
    const paths = args.filter((a) => !a.startsWith('-'));
    const summarize = (text, label) => {
      const lines = text === '' ? 0 : text.replace(/\n$/, '').split('\n').length;
      const words = text.trim() ? text.trim().split(/\s+/).length : 0;
      const bytes = text.length;
      return `${String(lines).padStart(7)} ${String(words).padStart(7)} ${String(bytes).padStart(7)}` + (label ? ` ${label}` : '');
    };
    if (!paths.length) return this._ok(summarize(stdin || ''));
    const lines = [];
    for (const p of paths) {
      const got = this._readFile(p);
      if (got.error) return this._err(got.error.replace(/^cat/, 'wc'));
      lines.push(summarize(got.text, p));
    }
    return this._ok(lines.join('\n'));
  },

  _cmdSort(args, stdin) {
    let text = stdin;
    const paths = args.filter((a) => !a.startsWith('-'));
    if (paths.length) {
      const got = this._readFile(paths[0]);
      if (got.error) return this._err(got.error.replace(/^cat/, 'sort'));
      text = got.text;
    }
    const lines = String(text || '').replace(/\n$/, '').split('\n');
    lines.sort();
    return this._ok(lines.join('\n'));
  },

  _cmdUniq(args, stdin) {
    let text = stdin;
    const paths = args.filter((a) => !a.startsWith('-'));
    if (paths.length) {
      const got = this._readFile(paths[0]);
      if (got.error) return this._err(got.error.replace(/^cat/, 'uniq'));
      text = got.text;
    }
    const lines = String(text || '').replace(/\n$/, '').split('\n');
    const out = [];
    for (const line of lines) {
      if (out[out.length - 1] !== line) out.push(line);
    }
    return this._ok(out.join('\n'));
  },

  _cmdGrep(args, stdin) {
    const flags = new Set();
    const rest = [];
    args.forEach((a) => {
      if (a.startsWith('-') && a !== '-') a.slice(1).split('').forEach((c) => flags.add(c));
      else rest.push(a);
    });
    if (!rest.length && !stdin) return this._err('Usage: grep [ -i ] [ -r ] PATTERN [FILE...]');
    const pattern = rest[0];
    const files = rest.slice(1);
    if (!files.length && !stdin && !flags.has('r') && !flags.has('R')) {
      return this._err('Usage: grep [ -i ] [ -r ] PATTERN [FILE...]');
    }
    let re;
    try {
      re = new RegExp(pattern, flags.has('i') ? 'i' : '');
    } catch {
      return this._err(`grep: invalid pattern`);
    }

    const hits = [];
    const searchText = (text, label) => {
      String(text).split('\n').forEach((line) => {
        if (re.test(line)) hits.push(label ? `${label}:${line}` : line);
      });
    };

    if (!files.length) {
      if (flags.has('r') || flags.has('R')) {
        files.push('.');
      } else {
        searchText(stdin || '', '');
        if (this.ctx && hits.length) this.ctx.grepHits = (this.ctx.grepHits || 0) + hits.length;
        return this._ok(hits.join('\n'));
      }
    }

    const visit = (pathArg) => {
      const res = VFS.resolve(this.vfs, this.cwd, pathArg, this.home);
      if (!res) {
        hits.push(`grep: ${pathArg}: No such file or directory`);
        return;
      }
      if (res.node.type === 'dir') {
        if (flags.has('r') || flags.has('R')) {
          VFS.walk(res.node, res.path, (node, path) => {
            if (node.type === 'file' && VFS.readable(node)) {
              const before = hits.length;
              searchText(node.content, path);
              if (hits.length > before && this.ctx) {
                this.ctx.readFiles = this.ctx.readFiles || [];
                if (!this.ctx.readFiles.includes(path)) this.ctx.readFiles.push(path);
              }
            }
          });
        } else {
          hits.push(`grep: ${pathArg}: Is a directory`);
        }
        return;
      }
      if (!VFS.readable(res.node)) {
        hits.push(`grep: ${pathArg}: Permission denied`);
        return;
      }
      if (this.ctx) {
        this.ctx.readFiles = this.ctx.readFiles || [];
        if (!this.ctx.readFiles.includes(res.path)) this.ctx.readFiles.push(res.path);
      }
      searchText(res.node.content, files.length + (flags.has('r') ? 1 : 0) > 1 ? res.path : '');
    };
    files.forEach(visit);
    if (this.ctx && hits.length) {
      this.ctx.grepHits = (this.ctx.grepHits || 0) + hits.length;
      this.ctx.grepPatterns = this.ctx.grepPatterns || [];
      this.ctx.grepPatterns.push(pattern.toLowerCase());
    }
    return this._ok(hits.join('\n'));
  }
});

Terminal.define('cat', function (args, stdin) { return this._cmdCat(args, stdin); });
Terminal.define(['less', 'more'], function (args, stdin) { return this._cmdLess(args, stdin); });
Terminal.define('head', function (args, stdin) { return this._cmdHeadTail(args, stdin, 'head'); });
Terminal.define('tail', function (args, stdin) { return this._cmdHeadTail(args, stdin, 'tail'); });
Terminal.define('wc', function (args, stdin) { return this._cmdWc(args, stdin); });
Terminal.define('sort', function (args, stdin) { return this._cmdSort(args, stdin); });
Terminal.define('uniq', function (args, stdin) { return this._cmdUniq(args, stdin); });
Terminal.define('grep', function (args, stdin) { return this._cmdGrep(args, stdin); });

File: crates/roc-kernel/src/shell/commands/text.rs
//! Text commands + headless pager. Matches `js/commands/text.js`.

use regex::RegexBuilder;

use crate::shell::exec::{retarget_cat, Pager};
use crate::vfs::{self, js_len};

use super::super::exec::Shell;
use super::super::output::Output;

impl Shell {
    pub(crate) fn cmd_cat(&mut self, args: &[String], stdin: &str) -> Output {
        if args.is_empty() {
            return Output::ok(stdin);
        }
        let mut chunks = Vec::new();
        for a in args {
            match self.read_file(a) {
                Ok(got) => {
                    self.note_read(&got.path);
                    chunks.push(got.text);
                }
                Err(e) => return Output::err(e),
            }
        }
        Output::ok(chunks.join(""))
    }

    pub(crate) fn cmd_less(&mut self, args: &[String], stdin: &str) -> Output {
        let paths: Vec<&str> = args
            .iter()
            .map(String::as_str)
            .filter(|a| !a.starts_with('-'))
            .collect();
        let (text, title) = if !paths.is_empty() {
            let mut chunks = Vec::new();
            let mut title = String::new();
            for a in &paths {
                match self.read_file(a) {
                    Ok(got) => {
                        self.note_read(&got.path);
                        chunks.push(got.text);
                        title = got.path;
                    }
                    Err(e) => return Output::err(e.replacen("cat:", "less:", 1)),
                }
            }
            let text = chunks.join("\n");
            if paths.len() > 1 {
                title = format!("{} files", paths.len());
            }
            (text, title)
        } else {
            let text = stdin.to_string();
            if text.is_empty() {
                return Output::err("less: missing filename (try: less FILE)");
            }
            (text, "(stdin)".to_string())
        };
        let lines: Vec<String> = split_lines_one_nl(&text)
            .into_iter()
            .map(str::to_string)
            .collect();
        self.pager = Some(Pager {
            lines,
            pos: 0,
            title,
            page_size: 20,
        });
        // JS returns empty stdout; the pager paints the screen.
        Output::ok("")
    }

    pub(crate) fn cmd_head_tail(&mut self, args: &[String], stdin: &str, kind: &str) -> Output {
        let mut n = 10usize;
        let mut paths = Vec::new();
        let mut i = 0;
        while i < args.len() {
            if args[i] == "-n" && i + 1 < args.len() {
                n = parse_count(&args[i + 1]);
                i += 2;
            } else if is_count_flag(&args[i]) {
                let digits = args[i].trim_start_matches("-n").trim_start_matches('-');
                n = parse_count(digits);
                i += 1;
            } else {
                paths.push(args[i].as_str());
                i += 1;
            }
        }
        let mut text = stdin.to_string();
        if let Some(path) = paths.first() {
            match self.read_file(path) {
                Ok(got) => {
                    self.note_read(&got.path);
                    text = got.text;
                }
                Err(e) => return Output::err(retarget_cat(&e, kind)),
            }
        }
        let lines = split_lines_one_nl(&text);
        let slice: Vec<&str> = if kind == "head" {
            lines.into_iter().take(n).collect()
        } else {
            let start = lines.len().saturating_sub(n);
            lines.into_iter().skip(start).collect()
        };
        Output::ok(slice.join("\n"))
    }

    pub(crate) fn cmd_wc(&mut self, args: &[String], stdin: &str) -> Output {
        let paths: Vec<&str> = args
            .iter()
            .map(String::as_str)
            .filter(|a| !a.starts_with('-'))
            .collect();
        if paths.is_empty() {
            return Output::ok(wc_line(stdin, None));
        }
        let mut rows = Vec::new();
        for p in paths {
            match self.read_file(p) {
                Ok(got) => rows.push(wc_line(&got.text, Some(p))),
                Err(e) => return Output::err(retarget_cat(&e, "wc")),
            }
        }
        Output::ok(rows.join("\n"))
    }

    pub(crate) fn cmd_sort(&mut self, args: &[String], stdin: &str) -> Output {
        let paths: Vec<&str> = args
            .iter()
            .map(String::as_str)
            .filter(|a| !a.starts_with('-'))
            .collect();
        let mut text = stdin.to_string();
        if let Some(path) = paths.first() {
            match self.read_file(path) {
                Ok(got) => text = got.text,
                Err(e) => return Output::err(retarget_cat(&e, "sort")),
            }
        }
        let mut lines: Vec<&str> = split_lines_one_nl(&text);
        lines.sort();
        Output::ok(lines.join("\n"))
    }

    pub(crate) fn cmd_uniq(&mut self, args: &[String], stdin: &str) -> Output {
        let paths: Vec<&str> = args
            .iter()
            .map(String::as_str)
            .filter(|a| !a.starts_with('-'))
            .collect();
        let mut text = stdin.to_string();
        if let Some(path) = paths.first() {
            match self.read_file(path) {
                Ok(got) => text = got.text,
                Err(e) => return Output::err(retarget_cat(&e, "uniq")),
            }
        }
        let lines = split_lines_one_nl(&text);
        let mut out: Vec<&str> = Vec::new();
        for line in lines {
            if out.last() != Some(&line) {
                out.push(line);
            }
        }
        Output::ok(out.join("\n"))
    }

    pub(crate) fn cmd_grep(&mut self, args: &[String], stdin: &str) -> Output {
        let mut flags = std::collections::HashSet::new();
        let mut rest = Vec::new();
        for a in args {
            if a.starts_with('-') && a != "-" {
                for c in a.chars().skip(1) {
                    flags.insert(c);
                }
            } else {
                rest.push(a.as_str());
            }
        }
        let recursive = flags.contains(&'r') || flags.contains(&'R');
        if rest.is_empty() && stdin.is_empty() {
            return Output::err("Usage: grep [ -i ] [ -r ] PATTERN [FILE...]");
        }
        if rest.is_empty() {
            return Output::err("Usage: grep [ -i ] [ -r ] PATTERN [FILE...]");
        }
        let pattern = rest[0];
        let mut files: Vec<String> = rest[1..].iter().map(|s| (*s).to_string()).collect();
        if files.is_empty() && stdin.is_empty() && !recursive {
            return Output::err("Usage: grep [ -i ] [ -r ] PATTERN [FILE...]");
        }
        let re = match RegexBuilder::new(pattern)
            .case_insensitive(flags.contains(&'i'))
            .build()
        {
            Ok(re) => re,
            Err(_) => return Output::err("grep: invalid pattern"),
        };

        let mut hits = Vec::new();
        let search_text = |text: &str, label: &str, hits: &mut Vec<String>| {
            for line in text.split('\n') {
                if re.is_match(line) {
                    if label.is_empty() {
                        hits.push(line.to_string());
                    } else {
                        hits.push(format!("{label}:{line}"));
                    }
                }
            }
        };

        if files.is_empty() {
            if recursive {
                files.push(".".to_string());
            } else {
                search_text(stdin, "", &mut hits);
                if !hits.is_empty() {
                    self.ctx.grep_hits += hits.len() as i32;
                }
                return Output::ok(hits.join("\n"));
            }
        }

        let label_paths = files.len() + usize::from(recursive) > 1;
        let mut newly_read = Vec::new();
        for path_arg in &files {
            match vfs::resolve(&self.vfs, &self.cwd, path_arg, &self.home) {
                None => hits.push(format!("grep: {path_arg}: No such file or directory")),
                Some(res) if res.node.is_dir() => {
                    if recursive {
                        vfs::walk(res.node, &res.path, |node, path| {
                            if node.is_file() && vfs::readable(node) {
                                let before = hits.len();
                                search_text(node.content().unwrap_or(""), path, &mut hits);
                                if hits.len() > before {
                                    newly_read.push(path.to_string());
                                }
                            }
                        });
                    } else {
                        hits.push(format!("grep: {path_arg}: Is a directory"));
                    }
                }
                Some(res) if !vfs::readable(res.node) => {
                    hits.push(format!("grep: {path_arg}: Permission denied"));
                }
                Some(res) => {
                    newly_read.push(res.path.clone());
                    let label = if label_paths { res.path.as_str() } else { "" };
                    search_text(res.node.content().unwrap_or(""), label, &mut hits);
                }
            }
        }
        for p in newly_read {
            self.note_read(&p);
        }
        if !hits.is_empty() {
            self.ctx.grep_hits += hits.len() as i32;
            self.ctx.grep_patterns.push(pattern.to_lowercase());
        }
        Output::ok(hits.join("\n"))
    }
}

fn split_lines_one_nl(text: &str) -> Vec<&str> {
    let t = text.strip_suffix('\n').unwrap_or(text);
    t.split('\n').collect()
}

fn is_count_flag(arg: &str) -> bool {
    // JS `/^-n?\d+$/`
    let Some(rest) = arg.strip_prefix('-') else {
        return false;
    };
    let rest = rest.strip_prefix('n').unwrap_or(rest);
    !rest.is_empty() && rest.bytes().all(|b| b.is_ascii_digit())
}

fn parse_count(s: &str) -> usize {
    match s.parse::<usize>() {
        Ok(0) | Err(_) => 10,
        Ok(n) => n,
    }
}

fn wc_line(text: &str, label: Option<&str>) -> String {
    let lines = if text.is_empty() {
        0
    } else {
        split_lines_one_nl(text).len()
    };
    let words = if text.trim().is_empty() {
        0
    } else {
        text.trim().split_whitespace().count()
    };
    let bytes = js_len(text);
    let row = format!("{lines:>7} {words:>7} {bytes:>7}");
    match label {
        Some(p) => format!("{row} {p}"),
        None => row,
    }
}
